csharp/src/lesson03/README.md

```diff
@@ -0,0 +1,327 @@
+
+public FormatController(ITracer tracer)
+{
+    _tracer = tracer;
+}
+```
+
+#### Extract the span context from the incoming request using `ITracer.Extract`
+
+First, add a helper function:
+
+```csharp
+public static IScope StartServerSpan(ITracer tracer, IDictionary<string, string> headers, string operationName)
+{
+    ISpanBuilder spanBuilder;
+    try
+    {
+        ISpanContext parentSpanCtx = tracer.Extract(BuiltinFormats.HttpHeaders, new TextMapExtractAdapter(headers));
+
+        spanBuilder = tracer.BuildSpan(operationName);
+        if (parentSpanCtx != null)
+        {
+            spanBuilder = spanBuilder.AsChildOf(parentSpanCtx);
+        }
+    }
+    catch (Exception)
+    {
+        spanBuilder = tracer.BuildSpan(operationName);
+    }
+
+    // TODO could add more tags like http.url
+    return spanBuilder.WithTag(Tags.SpanKind, Tags.SpanKindServer).StartActive(true);
+}
+```
+
+The logic here is similar to the client side instrumentation, except that we are using `_tracer.Extract`
+and tagging the span as `span.kind=server`. We are extracting the headers via `TextMapExtractAdapter`.
+
+Now change the `/api/format/helloString` handler method to use `StartServerSpan`:
+
+```csharp
+[HttpGet("{helloString}", Name = "GetFormat")]
+public string Get(string helloString)
+{
+    var headers = Request.Headers.ToDictionary(k => k.Key, v => v.Value.First());
+    using (var scope = StartServerSpan(_tracer, headers, "format-controller"))
+    {
+        var formattedHelloString = $"Hello, {helloString}!";
+        scope.Span.Log(new Dictionary<string, object>
+        {
+            [LogFields.Event] = "string-format",
+            ["value"] = formattedHelloString
+        });
+        return formattedHelloString;
+    }
+}
+```
+
+### Take It For a Spin
+
+As before, first run the `Lesson03.Exercise.Server` project in Visual Studio.
+Then run `Lesson03.Exercise.Client` in the terminal. You should see an output like this:
+
+Server output (ASP.NET Core Web Server):
+
+```powershell
+$ dotnet run
+info: Jaeger.Configuration[0]
+      Initialized Jaeger.Tracer
+Hosting environment: Development
+Content root path: opentracing-tutorial\csharp\src\lesson03\example\Lesson03.Example.Server
+Now listening on: http://localhost:8081
+Application started. Press Ctrl+C to shut down.
+info: Jaeger.Reporters.LoggingReporter[0]
+      Span reported: 722582a9299820e258eee39b348263cc:1924ee38b5e4b7c3:52a7512a6a2d7ccc:1 - format-controller
+info: Jaeger.Reporters.LoggingReporter[0]
+      Span reported: 722582a9299820e258eee39b348263cc:291f14c7a20258ba:52a7512a6a2d7ccc:1 - Result ObjectResult
+info: Jaeger.Reporters.LoggingReporter[0]
+      Span reported: 722582a9299820e258eee39b348263cc:52a7512a6a2d7ccc:1e05b0ca172ad56e:1 - Action OpenTracing.Tutorial.Lesson03.Example.Server.Controllers.FormatController/Get
+info: Jaeger.Reporters.LoggingReporter[0]
+      Span reported: 722582a9299820e258eee39b348263cc:1e05b0ca172ad56e:92ffe8a50d62d2e:1 - HTTP GET
```

Comment:
I would expect only two spans on the server, called `format-string` and `print-hello`.

Reply:
The two extras come from OpenTracing.Contrib.NetCore. I made the logs with it :) nice catch

Reply:
It will be only one span, since we only use `format-string` on the server side (as child with operation name `format-controller`) and do the `print-hello` only in the client to keep the example easier. I know that Java had two API endpoints (`format-string` and `print-hello`) but I dropped it since it doesn't bring any new insights to speed up the tutorial a bit.